### PR TITLE
apply NEP29

### LINKED ISSUE #1518: Adopt NEP29: minimum dependency policy
After Cartopy 0.18, we will be dropping support for Python 2. Currently, for Python 3, we have a minimum of 3.5. This is somewhat problematic on conda-forge because they've already dropped support for it.

For the future, we should implement something like [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) as a minimum dependency policy. For Python, this is a 42-month window and for NumPy, it is a 24-month window.

Assuming that 0.18 is out by the end of the month, and 0.19 takes at least 3 months, that'd be end of July. Extending that policy to other dependencies using 42 months (January 30, 2017), that would mean:
* Python 3.7
* NumPy 1.16
* Cython 0.26 (older than current requirement)
* Proj 5.0.0
* GEOS 3.6.2
* Shapely 1.6b4
* pyshp 1.2.11 (older than current requirement)
* pyepsg 0.3.2 (older than current requirement)
* OWSLib 0.15.0
* Pillow 4.1.0
* Matplotlib 2.0.1
* GDAL 2.2.0
* SciPy 0.19.0
* pytest 3.0.7 (older than current requirement)

Extending that policy to other dependencies using 24 months (July 30, 2018) instead would mean:
* Python 3.7
* NumPy 1.16
* Cython 0.28.5
* Proj 5.2.0
* GEOS 3.6.3
* Shapely 1.7a1
* pyshp 2.0.0
* pyepsg 0.4.0
* OWSLib 0.17.0
* Pillow 5.3.0
* Matplotlib 2.2.3
* GDAL 2.3.2
* SciPy 1.2.0
* pytest 3.7.0

Though note, we probably shouldn't switch minimums to alphas or betas; that's just there for reference. We may also want slightly different policy for system deps like Proj and GEOS.

### PR DESCRIPTION
**Pull request by graingert**
Fixes #1518

### CODE REVIEW

**Comment by graingert** [comment]
Looks like the image changed with v3.7 but I don't have a Mac to generate it on

**Comment by greglucas** [comment]
As a general comment, I think this is a good thing, but that it should wait until after the 0.19 release to go in. It'd be nice to keep the lower Python support on 0.19 since that will be our first release to officially drop Python 2 and make it an easier transition if people are still running Python 3.6.

The matplotlib `conda` is downloading isn't linked to the proper freetype, which I think may be what is actually causing the image failure (MPL should be downloaded from the label/testing channel). However, I'm not sure why the `Upload Images` step isn't getting run... I put that in there so we could easily get the image results out of the CI system to re-use in situations like this. I swear I tested it on my fork and it worked for me, so I'm not sure what changed.

Specifically, this isn't getting called: https://github.com/SciTools/cartopy/blob/7237260b04f6d803c4b0f0f6095a27a4f83adaef/.github/workflows/ci-testing.yml#L81
So maybe the `test` step conclusion isn't `failure`, or it fails fast on this job? No need for you to look into this one if you don't want, just putting this down for future reference.

**Comment by greglucas** [comment]
@graingert, v0.19 is out now, so I think this would be OK to put in now and drop some older support dependencies. Can you rebase and we'll see if the macos build is still failing?

**Comment by QuLogic** [comment]
It looks like there are still conflicts?

File: setup.py [hunk 1]
```diff
@@ -179,7 +179,7 @@ def get_proj_libraries():
     proj_libraries = ["proj"]
     if os.name == "nt" and (6, 0, 0) <= proj_version < (6, 3, 0):
         proj_libraries = [
-            "proj_{}_{}".format(proj_version[0], proj_version[1])
+            f"proj_{proj_version[0]}_{proj_version[1]}"
```
Comment (QuLogic) [inline setup.py:182]: You can't make this change to `setup.py`; it needs to remain Python 2 compatible.

Comment (graingert) [reply setup.py:182]: No it doesn't, the requires_python will prevent pip from downloading the sdist

Comment (QuLogic) [reply setup.py:182]: It does; read the comment at the top.

**Comment by greglucas** [comment]
Updated and incorporated in https://github.com/SciTools/cartopy/pull/1844

**Pull request closed without merging**